bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -126,26 +127,34 @@ impl AuthenticationService {
     ) -> Result<(), AuthenticationError> {
         let mut builder = Arc::new(ClientBuilder::new()).base_path(self.base_path.clone());
 
-        // Remove any URL scheme from the name to attempt discovery first.
-        let server_name = matrix_sdk::sanitize_server_name(&server_name_or_homeserver_url)
-            .map_err(AuthenticationError::from)?;
-
-        builder = builder.server_name(server_name.to_string());
-
-        let client = builder
-            .build()
-            .or_else(|e| {
-                if !server_name_or_homeserver_url.starts_with("http://")
-                    && !server_name_or_homeserver_url.starts_with("http://")
-                {
-                    return Err(e);
+        // Attempt discovery as a server name first.
+        let result = matrix_sdk::sanitize_server_name(&server_name_or_homeserver_url);
+        match result {
+            Ok(server_name) => {
+                builder = builder.server_name(server_name.to_string());
+            }
+            Err(e) => {
+                // When the input isn't a valid server name check it is a URL.
+                // If this is the case, build the client with a homeserver URL.
+                if let Ok(_url) = Url::parse(&server_name_or_homeserver_url) {
+                    builder = builder.homeserver_url(server_name_or_homeserver_url.clone());
```

Comment:
I think this shouldn't just be a regular fallback here, but rather integrated with the UI in some way (I think users should get some sort of warning that tells them about broken discovery so they can let their HS admin know).